### PR TITLE
Document Hermes' gRPC endpoint requirements

### PR DESCRIPTION
**Pull request by seanchen1991**
Closes: #3545 

## Description

______

### PR author checklist:
- [ ] Added changelog entry, using [`unclog`](https://github.com/informalsystems/unclog).
- [ ] Added tests: integration (for Hermes) or unit/mock tests (for modules).
- [ ] Linked to GitHub issue.
- [ ] Updated code comments and documentation (e.g., `docs/`).
- [ ] Tagged *one* reviewer who will be the one responsible for shepherding this PR.

### Reviewer checklist:

- [ ] Reviewed `Files changed` in the GitHub PR explorer.
- [ ] Manually tested (in case integration/unit/mock tests are absent).

### CODE REVIEW

**Comment by romac** [comment]
While we are at it, we should also document all the gRPC queries Hermes makes. Or should we do it in another PR?

**Comment by seanchen1991** [comment]
Yea, we can document the gRPC queries as well as part of this PR.

File: docs/spec/relayer/Data-Requirements.md [hunk 1]
```diff
@@ -3,17 +3,17 @@
...
+  * [`/status`](#status)
+  * [`/header`](#header)
+  * [`/latest_commit`, `/commit`, `/validators`](#latest_commit-commit-validators)
+  * [`/abci_query`](#abci_query)
+  * [`/tx_search`](#tx_search)
+  * [`/block_search`](#block_search)
+  * [`/block_results`](#block_results)
+  * [`/broadcast_tx_sync`](#broadcast_tx_sync)
+  * [`/broadcast_evidence`](#broadcast_evidence)
 - [CometBFT WebSocket](#cometbft-websocket)
```
Comment (romac) [inline docs/spec/relayer/Data-Requirements.md:17]:
```suggestion
- [CometBFT WebSocket](#cometbft-websocket)
- [CometBFT gRPC](#cometbft-grpc)
```

File: docs/spec/relayer/Data-Requirements.md [hunk 2]
```diff
@@ -162,3 +162,30 @@ The relayer connects to the node's CometBFT websocket interface and subscribes t
     - `EXISTS acknowledge_packet`
     - `EXISTS timeout_packet`
 
+## CometBFT gRPC
+
+The following are the gRPC endpoints that Hermes makes use of, listed in rough priority from most used to least used. 
+
+### `ibc.core.channel.v1.QueryClient`
```
Comment (romac) [inline docs/spec/relayer/Data-Requirements.md:169]: On top of listing the client/service used to perform the query, can we also list the actual queries that are being made?

eg. for `ibc.core.channel.v1.QueryClient` that would be:
- `QueryChannelClientStateRequest`
- `QueryChannelsRequest`
- `QueryConnectionChannelsRequest`
- `QueryNextSequenceReceiveRequest`
- `QueryPacketAcknowledgementsRequest`
- `QueryPacketCommitmentsRequest`
- `QueryUnreceivedAcksRequest`
- `QueryUnreceivedPacketsRequest`

File: docs/spec/relayer/Data-Requirements.md [hunk 3]
```diff
@@ -162,3 +163,54 @@ The relayer connects to the node's CometBFT websocket interface and subscribes t
...
+
+### `interchain_security.ccv.consumer.v1.QueryClient`
+
+Queries a CCV Consumer chain to fetch its staking parameters, notably its unbonding period and the number of historical entries that the chain keeps.
+
+### `cosmos.staking.v1beta1.QueryClient`
+
+Queries a Cosmos chain to fetch its staking parameters, most notably the chain's unbonding period.
+
+### `cosmos.base.node.v1beta1.ServiceClient`
```
Comment (romac) [inline docs/spec/relayer/Data-Requirements.md:214]: Can we list the specific queries we make using this client?

File: docs/spec/relayer/Data-Requirements.md [hunk 4]
```diff
@@ -162,3 +163,54 @@ The relayer connects to the node's CometBFT websocket interface and subscribes t
...
+The requests that Hermes makes to this endpoint are:
+
+- QueryClientConnectionsRequest: Requests all connections associated with a specified client.
+- QueryConnectionsRequest: Requests all connections associated with the chain.
+
+### `interchain_security.ccv.consumer.v1.QueryClient`
+
+Queries a CCV Consumer chain to fetch its staking parameters, notably its unbonding period and the number of historical entries that the chain keeps.
+
+### `cosmos.staking.v1beta1.QueryClient`
```
Comment (romac) [inline docs/spec/relayer/Data-Requirements.md:210]: Can we list the specific queries we make using this client?

File: docs/spec/relayer/Data-Requirements.md [hunk 5]
```diff
@@ -162,3 +163,54 @@ The relayer connects to the node's CometBFT websocket interface and subscribes t
...
+### `ibc.core.connection.v1.QueryClient`
+
+Queries the connection in order to fetch connection data and connected clients.
+
+The requests that Hermes makes to this endpoint are:
+
+- QueryClientConnectionsRequest: Requests all connections associated with a specified client.
+- QueryConnectionsRequest: Requests all connections associated with the chain.
+
+### `interchain_security.ccv.consumer.v1.QueryClient`
```
Comment (romac) [inline docs/spec/relayer/Data-Requirements.md:206]: Can we list the specific queries we make using this client?

File: docs/spec/relayer/Data-Requirements.md [hunk 6]
```diff
@@ -162,3 +163,54 @@ The relayer connects to the node's CometBFT websocket interface and subscribes t
...
+- QueryConsensusStateHeightsRequest: Requests all the consensus state heights associated with a specified client.
+- QueryConsensusStatesRequest: Requests all the consensus states associated with a specified client.
+
+### `ibc.core.connection.v1.QueryClient`
+
+Queries the connection in order to fetch connection data and connected clients.
+
+The requests that Hermes makes to this endpoint are:
+
+- QueryClientConnectionsRequest: Requests all connections associated with a specified client.
```
Comment (romac) [inline docs/spec/relayer/Data-Requirements.md:203]:
```suggestion
- `QueryClientConnections`: Requests all connections associated with a specified client.
```

File: docs/spec/relayer/Data-Requirements.md [hunk 7]
```diff
@@ -162,3 +163,54 @@ The relayer connects to the node's CometBFT websocket interface and subscribes t
...
+- QueryConnectionsRequest: Requests all connections associated with the chain.
```
Comment (romac) [inline docs/spec/relayer/Data-Requirements.md:204]:
```suggestion
- `QueryConnections`: Requests all connections associated with the chain.
```

File: docs/spec/relayer/Data-Requirements.md [hunk 8]
```diff
@@ -162,3 +163,54 @@ The relayer connects to the node's CometBFT websocket interface and subscribes t
...
+
+### `ibc.core.client.v1.QueryClient`
+
+Queries the client in order to fetch client and consensus states.
+
+The requests that Hermes makes to this endpoint are:
+
+- QueryClientStatesRequest: Requests all client states associated with the chain.
+- QueryConsensusStateHeightsRequest: Requests all the consensus state heights associated with a specified client.
+- QueryConsensusStatesRequest: Requests all the consensus states associated with a specified client.
```
Comment (romac) [inline docs/spec/relayer/Data-Requirements.md:195]:
```suggestion
- `QueryClientStates`: Requests all client states associated with the chain.
- `QueryConsensusStateHeights`: Requests all the consensus state heights associated with a specified client.
- `QueryConsensusStates`: Requests all the consensus states associated with a specified client.
```

File: docs/spec/relayer/Data-Requirements.md [hunk 9]
```diff
@@ -162,3 +163,54 @@ The relayer connects to the node's CometBFT websocket interface and subscribes t
...
+- QueryChannelsRequest: Requests all of the channels associated with the chain.
+- QueryConnectionChannelsRequest: Requests all of the channels associated with a specified connection.
+- QueryNextSequenceReceiveRequest: Requests the sequence number of the next receive packet for a specified channel.
+- QueryPacketCommitmentsRequest: Requests the packet commitments associated with a specified channel.
+- QueryPacketAcknowledgementsRequest: Requests the packet acknowledgments associated with a specified channel.
+- QueryUnreceivedAcksRequest: Requests the unreceived acknowledgments associated with a specified channel.
+- QueryUnreceivedPacketsRequest: Requests the unreceived packet sequences associated with a specified channel.
+
+> _*Note:*_ The `QueryPacketAcknowledgementsRequest`, `QueryUnreceivedAcksRequest`, and `QueryUnreceivedPacketsRequest` types each accept a Vector of `sequence`s in order to specify which packet commitments to fetch acknowledgements for. In the case where no `sequence`s are passed, the requests will simply return all acknowledgements for all outstanding packet commitments.
```
Comment (romac) [inline docs/spec/relayer/Data-Requirements.md:185]:
```suggestion
- `QueryChannelClientState`: Requests the client state associated with a specified channel.
- `QueryChannels`: Requests all of the channels associated with the chain.
- `QueryConnectionChannels`: Requests all of the channels associated with a specified connection.
- `QueryNextSequenceReceive`: Requests the sequence number of the next receive packet for a specified channel.
- `QueryPacketCommitments`: Requests the packet commitments associated with a specified channel.
- `QueryPacketAcknowledgements`: Requests the packet acknowledgments associated with a specified channel.
- `QueryUnreceivedAcks`: Requests the unreceived acknowledgments associated with a specified channel.
- `QueryUnreceivedPackets`: Requests the unreceived packet sequences associated with a specified channel.

> _*Note:*_ The `QueryPacketAcknowledgements`, `QueryUnreceivedAcks`, and `QueryUnreceivedPackets` queries each accept a vector of `sequences` in order to specify which packet commitments to fetch acknowledgements for. In the case where an empty vector is passed, the queries will simply return all acknowledgements for all outstanding packet commitments. 
```

File: docs/spec/relayer/Data-Requirements.md (hunk 3 above)
Comment (seanchen1991) [reply docs/spec/relayer/Data-Requirements.md:220]: [dfe0c6c](https://github.com/informalsystems/hermes/pull/3529/commits/dfe0c6cbe566553a4f1c85c73087e541ce22c97c)

File: docs/spec/relayer/Data-Requirements.md (hunk 5 above)
Comment (seanchen1991) [reply docs/spec/relayer/Data-Requirements.md:206]: [dfe0c6c](https://github.com/informalsystems/hermes/pull/3529/commits/dfe0c6cbe566553a4f1c85c73087e541ce22c97c)

**Pull request merged**